sdks/python/apache_beam/io/gcsio.py

```diff
@@ -539,7 +546,41 @@ def _fetch_next_if_buffer_exhausted(self):
         self.buffer_start_position + len(self.buffer) <= self.position):
       bytes_to_request = min(self._remaining(), self.buffer_size)
       self.buffer_start_position = self.position
-      self.buffer = self._get_segment(self.position, bytes_to_request)
+      while True:
+        queue = Queue.Queue()
+        t = threading.Thread(target=self._fetch_to_queue,
+                             args=(queue, self._get_segment,
+                                   (self.position, bytes_to_request)))
+        t.daemon = True
+        t.start()
+        try:
+          result, exn = queue.get(timeout=self.segment_timeout)
+          if exn:
+            logging.error(
+                'Exception while fetching %d bytes from position %d of %s: %s.',
+                bytes_to_request, self.position, self.path, exn)
+            raise exn
+          self.buffer = result
+          return
+        except Queue.Empty:
+          logging.warning(
```

Comment:
OK. But maybe add a note there at least, that we are leaking a thread. They can potentially hang forever.